### PR TITLE
nspawn: do not uid shift mount points if --read-only

### PR DESCRIPTION
**Pull request by lotheac**
Hi, we're using systemd-nspawn to create single-use containers for software builds. We want to use a simple user mapping like '--private-users=$(id -u):1' so that container root can write to bind-mounted output paths from inside the container. This works well with --ephemeral, but unfortunately we can neither use btrfs nor copy the root filesystem every time: some of our container root filesystems are tens of gigabytes large, and we ran into memory exhaustion issues with btrfs cleaning up after creating and destroying several clones.

So, instead we'd like to use --read-only, but still use user namespaces. However, currently systemd-nspawn insists on chown()ing the (already existing, owned by host root) mount points (eg. container /tmp) to the container root user prior to mounting on them, which fails with EROFS if the container rootfs is read-only. The following patch skips chown() if the owner and group are already correct when creating the mount point, and skips applying uid shift to the mount point if --read-only and user namespaces are enabled.

### CODE REVIEW

**Comment by poettering** [comment]
hmm, could you include the rationale in the commit text itself, please?

**Comment by lotheac** [comment]
On Wed, Oct 18 2017 13:56:33 +0000, Lennart Poettering wrote:
> hmm, could you include the rationale in the commit text itself, please?

Commit message amended.

However, I've also found that there are certain new cases where we fail,
particularly when adding --private-network:

1. /sys is mounted as tmpfs in mount_all
2. another /sys mount is examined but not mounted "/* skipped if above
was mounted */"; however this causes mkdir_userns_p to be called again
3. mkdir_userns_p chown()s /sys (and its parents, which is relevant for
some other tmpfs mounts, like /dev and /dev/shm)
4. when mount_sysfs is called, /sys will be mounted tmpfs with the
correct uid and gid options, but will have been re-chowned to host root

I've added another commit to the PR to address this; it makes
--read-only --private-users=$(id -u):1 --private-network work also.

-- 
Lauri Tirkkonen | lotheac @ IRCnet

**Comment by poettering** [comment]
so, ultimately the ownership of the directory doesn't really matter, since we immediately overmount it. It's nice if it is properly owned, but it doesn't really matter, because it shouldn't be visible during runtime.

With your code you insist that the if --read-only is used that the dir is owned by UID 0. I am not sure we should do that though, simply because that would make it hard to switch between --read-only invocations and invocations without it.

I'd thus suggest a slightly different solution: let's change the mount_all() code to handle mkdir_userns_p() returning EROFS gracefully: just immediately proceed with the verbose_mount() invocation. This should be safe for all cases where EROFS is generated: 

1. when the dir to overmount doesn't exist and thus mkdir() already fails, then mount() will fail too with ENOENT, which should be discriptive enough

2. when the dir to overmount exists, and thus mkdir() fails with EEXIST and we then try to chmod() what is already there, then this will return EROFS, which we accept and proceed with the mount(), and the wrong ownership will not be visible

**Comment by poettering** [comment]
ah, uh, so i was writing the above rewview text at the time you pushed the update, so the text might not match entirely anymore what the commit contains, but i think it's still the way to go, i.e. just accept EROFS  gracefully, but not other errors

**Comment by lotheac** [comment]
On Wed, Oct 18 2017 16:01:05 +0000, Lennart Poettering wrote:
> ah, uh, so i was writing the above rewview text at the time you pushed the update, so the text might not match entirely anymore what the commit contains, but i think it's still the way to go, i.e. just accept EROFS  gracefully, but not other errors

You're right, that approach certainly makes the diff simpler. Updated.

-- 
Lauri Tirkkonen | lotheac @ IRCnet

File: src/nspawn/nspawn-mount.c [hunk 1]
```diff
@@ -597,7 +597,7 @@ int mount_all(const char *dest,
                         continue;
 
                 r = mkdir_userns_p(dest, where, 0755, mount_settings, uid_shift);
-                if (r < 0 && r != -EEXIST) {
+                if (r < 0 && r != -EEXIST && r != -EROFS) {
```
Comment (poettering) [inline src/nspawn/nspawn-mount.c]: hmm, so if you check r against to specific values please use !IN_SET(r, -EEXIST, -EROFS). 

that said, I think EROFS (unlike EEXIST) should result in a log message at LOG_DEBUG level. Even though it's safe to ignore it is useful to know that mkdir failed due to EROFS when trying to track down what goes on...

hence, I think this should be written as:

```c
…
if (r < 0 && r != -EEXIST) {
        log_errno_debug(fatal && r != -EROFS ? LOG_ERR : LOG_DEBUG, r, "Failed to create directory %s: %m", where);
        if (r != -EROFS) { /* if the root directory is read-only, and we couldn't create the directory hence then that's OK, let's continue, as mount() will then log as needed */
                if (fatal)
                        return r;
                continue;
        }
}
…
```

**Comment by lotheac** [comment]
On Wed, Oct 18 2017 18:36:17 +0000, Lennart Poettering wrote:
> > @@ -597,7 +597,7 @@ int mount_all(const char *dest,
>                          continue;
>  
>                  r = mkdir_userns_p(dest, where, 0755, mount_settings, uid_shift);
> -                if (r < 0 && r != -EEXIST) {
> +                if (r < 0 && r != -EEXIST && r != -EROFS) {
> 
> hmm, so if you check r against to specific values please use !IN_SET(r, -EEXIST, -EROFS). 
> 
> that said, I think EROFS (unlike EEXIST) should result in a log message at LOG_DEBUG level. Even though it's safe to ignore it is useful to know that mkdir failed due to EROFS when trying to track down what goes on...

Ok, updated commit.

> hence, I think this should be written as:
> 
> ```c
> …
> if (r < 0 && r != -EEXIST) {
>         log_errno_debug(fatal && r != -EROFS ? LOG_ERR : LOG_DEBUG, r, "Failed to create directory %s: %m", where);
>         if (r != -EROFS) { /* if the root directory is read-only, and we couldn't create the directory hence then that's OK, let's continue, as mount() will then log as needed */
>                 if (fatal)
>                         return r;
>                 continue;
>         }
> }
> …
> ```

I did it a little differently, but feel free to use your version if you
like it better.

-- 
Lauri Tirkkonen | lotheac @ IRCnet

**Comment by poettering** [comment]
lgtm, thanks!

**Comment by poettering** [comment]
hmm, any chance you can rebase this PR on current git and force push? The CI was broken with git until just now, and this was fixed a few moments ago with another commit. To make sure we don't break it right-away again, it would be great if you could rebase this PR and force push so that the CI runs over this again?

**Comment by lotheac** [comment]
On Tue, Oct 24 2017 07:43:54 +0000, Lennart Poettering wrote:
> hmm, any chance you can rebase this PR on current git and force push?

sure, done.

-- 
Lauri Tirkkonen | lotheac @ IRCnet

**Comment by poettering** [comment]
excellent! thanks!

**Pull request merged**